File: Classes/Utils/API/Clans/ClansWar.cpp
#include "Utils/API/Clans/ClansWar.h"

#include "Utils/API/URLEncoder.h"
#include "json/document.h"
#include "json/stringbuffer.h"
#include "json/writer.h"

void ClansWar::startWar(const std::string& clans_id,
                        StartWarCallback callback) {
  std::string url = BASE_URL + "/api/clanswar/start";
  url += "?clans_id=" + urlEncode(clans_id);

  auto request = new (std::nothrow) network::HttpRequest();
  if (!request) {
    if (callback) callback(false, "Failed to create HTTP request", "", "");
    return;
  }

  request->setUrl(url.c_str());
  request->setRequestType(network::HttpRequest::Type::GET);
  request->setResponseCallback(
      [callback](network::HttpClient* client, network::HttpResponse* response) {
        bool success = false;
        std::string message = "Unknown error";
        std::string war_id = "";
        std::string history_path = "";

        if (!response) {
          if (callback) callback(false, "Response is null", "", "");
          return;
        }

        int statusCode = response->getResponseCode();
        std::vector<char>* buffer = response->getResponseData();
        if (buffer && !buffer->empty()) {
          std::string resp(buffer->begin(), buffer->end());
          rapidjson::Document doc;
          doc.Parse(resp.c_str());
          if (!doc.HasParseError() && doc.IsObject()) {
            if (doc.HasMember("success") && doc["success"].IsBool())
              success = doc["success"].GetBool();
            if (doc.HasMember("message") && doc["message"].IsString())
              message = doc["message"].GetString();
            if (doc.HasMember("war_id") && doc["war_id"].IsString())
              war_id = doc["war_id"].GetString();
            if (doc.HasMember("history_path") && doc["history_path"].IsString())
              history_path = doc["history_path"].GetString();
          } else {
            message = "Failed to parse response";
          }
        } else {
          if (statusCode == 400)
            message = "Bad request";
          else
            message = "Empty response";
        }

        if (callback) callback(success, message, war_id, history_path);
      });

  network::HttpClient::getInstance()->send(request);
  request->release();
}

void ClansWar::getWarOverview(const std::string& clans_id,
                              WarOverviewCallback callback) {
  std::string url = BASE_URL + "/api/clanswar/overview";
  url += "?clans_id=" + urlEncode(clans_id);

  auto request = new (std::nothrow) network::HttpRequest();
  if (!request) {
    if (callback) {
      WarOverviewResult r;
      r.success = false;
      r.message = "Failed to create HTTP request";
      callback(r);
    }
    return;
  }

  request->setUrl(url.c_str());
  request->setRequestType(network::HttpRequest::Type::GET);
  request->setResponseCallback(
      [callback](network::HttpClient* client, network::HttpResponse* response) {
        WarOverviewResult result;
        if (!response) {
          result.success = false;
          result.message = "Response is null";
          if (callback) callback(result);
          return;
        }
        int statusCode = response->getResponseCode();
        std::vector<char>* buffer = response->getResponseData();
        if (buffer && !buffer->empty()) {
          std::string resp(buffer->begin(), buffer->end());
          rapidjson::Document doc;
          doc.Parse(resp.c_str());
          if (!doc.HasParseError() && doc.IsObject()) {
            if (doc.HasMember("success") && doc["success"].IsBool())
              result.success = doc["success"].GetBool();
            if (doc.HasMember("message") && doc["message"].IsString())
              result.message = doc["message"].GetString();
            if (doc.HasMember("overview") && doc["overview"].IsObject()) {
              const rapidjson::Value& ov = doc["overview"];
              for (auto itr = ov.MemberBegin(); itr != ov.MemberEnd(); ++itr) {
                WarMapOverview m;
                m.id = itr->name.GetString();
                const rapidjson::Value& v = itr->value;
                if (v.HasMember("stars") && v["stars"].IsInt())
                  m.stars = v["stars"].GetInt();
                if (v.HasMember("cnt") && v["cnt"].IsInt())
                  m.cnt = v["cnt"].GetInt();
                result.maps.push_back(m);
              }
            }
          } else {
            result.success = false;
            result.message = "Failed to parse response";
          }
        } else {
          result.success = false;
          if (statusCode == 400)
            result.message = "Bad request";
          else
            result.message = "Empty response";
        }
        if (callback) callback(result);
      });

  network::HttpClient::getInstance()->send(request);
  request->release();
}

void ClansWar::getWarMap(const std::string& clans_id, const std::string& map_id,
                         GetMapCallback callback) {
  std::string url = BASE_URL + "/api/clanswar/map";
  url += "?clans_id=" + urlEncode(clans_id);
  url += "&map_id=" + urlEncode(map_id);

  auto request = new (std::nothrow) network::HttpRequest();
  if (!request) {
    if (callback) callback(false, "Failed to create HTTP request", "");
    return;
  }

  request->setUrl(url.c_str());
  request->setRequestType(network::HttpRequest::Type::GET);
  request->setResponseCallback(
      [callback](network::HttpClient* client, network::HttpResponse* response) {
        bool success = false;
        std::string message = "Unknown error";
        std::string data = "";
        if (!response) {
          if (callback) callback(false, "Response is null", "");
          return;
        }
        int statusCode = response->getResponseCode();
        std::vector<char>* buffer = response->getResponseData();
        if (buffer && !buffer->empty()) {
          std::string resp(buffer->begin(), buffer->end());
          rapidjson::Document doc;
          doc.Parse(resp.c_str());
          if (!doc.HasParseError() && doc.IsObject()) {
            if (doc.HasMember("success") && doc["success"].IsBool())
              success = doc["success"].GetBool();
            if (doc.HasMember("message") && doc["message"].IsString())
              message = doc["message"].GetString();
            if (doc.HasMember("data")) {
              // 将 data 原样返回为字符串
              rapidjson::StringBuffer bufferOut;
              rapidjson::Writer<rapidjson::StringBuffer> writer(bufferOut);
              doc["data"].Accept(writer);
              data = bufferOut.GetString();
            }
          } else {
            message = "Failed to parse response";
          }
        } else {
          if (statusCode == 400)
            message = "Bad request";
          else
            message = "Empty response";
        }
        if (callback) callback(success, message, data);
      });

  network::HttpClient::getInstance()->send(request);
  request->release();
}

void ClansWar::saveWarMap(const std::string& clans_id,
                          const std::string& map_id,
                          const std::string& map_data_json,
                          SaveMapCallback callback) {
  std::string url = BASE_URL + "/api/clanswar/map/save";
  url += "?clans_id=" + urlEncode(clans_id);
  url += "&map_id=" + urlEncode(map_id);
  url += "&map_data=" + urlEncode(map_data_json);

  auto request = new (std::nothrow) network::HttpRequest();
  if (!request) {
    if (callback) callback(false, "Failed to create HTTP request");
    return;
  }

  request->setUrl(url.c_str());
  request->setRequestType(network::HttpRequest::Type::GET);
  request->setResponseCallback(
      [callback](network::HttpClient* client, network::HttpResponse* response) {
        bool success = false;
        std::string message = "Unknown error";
        if (!response) {
          if (callback) callback(false, "Response is null");
          return;
        }
        int statusCode = response->getResponseCode();
        std::vector<char>* buffer = response->getResponseData();
        if (buffer && !buffer->empty()) {
          std::string resp(buffer->begin(), buffer->end());
          rapidjson::Document doc;
          doc.Parse(resp.c_str());
          if (!doc.HasParseError() && doc.IsObject()) {
            if (doc.HasMember("success") && doc["success"].IsBool())
              success = doc["success"].GetBool();
            if (doc.HasMember("message") && doc["message"].IsString())
              message = doc["message"].GetString();
          } else {
            message = "Failed to parse response";
          }
        } else {
          if (statusCode == 400)
            message = "Bad request";
          else
            message = "Empty response";
        }
        if (callback) callback(success, message);
      });

  network::HttpClient::getInstance()->send(request);
  request->release();
}

File: server/app/api/clanswar.py
from flask import Blueprint, request, jsonify
import json
from app.utils.ClansWar import (
	start_war,
	get_war_overview,
	get_war_map,
	update_war_map,
)


clanswar_bp = Blueprint('clanswar', __name__)


@clanswar_bp.route('/clanswar/start', methods=['GET'])
def start():
	"""开启部落战
	GET 参数: clans_id
	返回: {success, message, war_id, history_path}
	"""
	clans_id = request.args.get('clans_id')
	if not clans_id:
		return jsonify({"success": False, "message": "clans_id 不能为空", "war_id": None, "history_path": None}), 400

	success, message, war_id, history_path = start_war(clans_id)
	return jsonify({"success": success, "message": message, "war_id": war_id, "history_path": history_path}), 200 if success else 400


@clanswar_bp.route('/clanswar/overview', methods=['GET'])
def overview():
	"""获取部落战概览
	GET 参数: clans_id
	返回: {success, message, overview}
	"""
	clans_id = request.args.get('clans_id')
	if not clans_id:
		return jsonify({"success": False, "message": "clans_id 不能为空", "overview": None}), 400

	success, message, overview = get_war_overview(clans_id)
	return jsonify({"success": success, "message": message, "overview": overview}), 200 if success else 404


@clanswar_bp.route('/clanswar/map', methods=['GET'])
def get_map():
	"""获取部落战某张地图
	GET 参数: clans_id, map_id
	返回: {success, message, data}
	"""
	clans_id = request.args.get('clans_id')
	map_id = request.args.get('map_id')
	if not clans_id or not map_id:
		return jsonify({"success": False, "message": "clans_id 和 map_id 不能为空", "data": None}), 400

	success, message, data = get_war_map(clans_id, map_id)
	return jsonify({"success": success, "message": message, "data": data}), 200 if success else 404


@clanswar_bp.route('/clanswar/map/save', methods=['GET'])
def save_map():
	"""保存/更新部落战地图
	GET 参数: clans_id, map_id, map_data (JSON字符串)
	返回: {success, message}
	"""
	clans_id = request.args.get('clans_id')
	map_id = request.args.get('map_id')
	map_data_str = request.args.get('map_data')

	if not clans_id or not map_id or not map_data_str:
		return jsonify({"success": False, "message": "clans_id、map_id 和 map_data 不能为空"}), 400

	try:
		map_data = json.loads(map_data_str)
	except json.JSONDecodeError as e:
		return jsonify({"success": False, "message": f"地图数据格式错误: {e}"}), 400

	success, message = update_war_map(clans_id, map_id, map_data)
	return jsonify({"success": success, "message": message}), 200 if success else 400
